pygmt/helpers/utils.py

```diff
@@ -61,23 +66,41 @@ def data_kind(data, x=None, y=None, z=None, required_z=False):
     'file'
     >>> data_kind(data=pathlib.Path("my-data-file.txt"), x=None, y=None)
     'file'
+    >>> data_kind(data=None, x=None, y=None, optional_data=True)
+    'null'
+    >>> data_kind(data=2.0, x=None, y=None, optional_data=True)
+    'null'
+    >>> data_kind(data=True, x=None, y=None, optional_data=True)
+    'null'
     >>> data_kind(data=xr.DataArray(np.random.rand(4, 3)))
     'grid'
     """
-    if data is None and x is None and y is None:
-        raise GMTInvalidInput("No input data provided.")
-    if data is not None and (x is not None or y is not None or z is not None):
-        raise GMTInvalidInput("Too much data. Use either data or x and y.")
-    if data is None and (x is None or y is None):
+    # pylint: disable=too-many-branches
+    # validate the combinations of data/x/y/z
```

Comment:
Kinda wishing that we can use switch statements here, but that's only Python 3.10 and above :upside_down_face: 

- https://stackoverflow.com/questions/11479816/what-is-the-python-equivalent-for-a-case-switch-statement
- https://docs.python.org/3.10/whatsnew/3.10.html#pep-634-structural-pattern-matching
- https://peps.python.org/pep-0634/

Reply:
Yes, we can revisit many if-else clauses after we bump to Python>=3.10.